File: searchin algorithms/binary_search_recursion.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int low, int high, int key) {
    if (low <= high) {
        int mid = low + (high - low) / 2;

        // Check if the key is present at mid
        if (arr[mid] == key)
            return mid;

        // If key is smaller than mid, it can only be in the left subarray
        if (arr[mid] > key)
            return binarySearch(arr, low, mid - 1, key);

        // Otherwise, the key can only be in the right subarray
        return binarySearch(arr, mid + 1, high, key);
    }

    // If the key is not present in the array
    return -1;
}

int main() {
    int arr[] = {2, 3, 4, 10, 40};
    int n = sizeof(arr) / sizeof(arr[0]);
    int key = 40;

    int result = binarySearch(arr, 0, n - 1, key);

    if (result != -1)
        cout << "Element found at index " << result << endl;
    else
        cout << "Element not found in the array" << endl;

    return 0;
}

File: searchin algorithms/binary_search_recursion.py
def binary_search(arr,left,right,key):
    if left > right:
        return -1 # target not found 
    mid = (left + right)//2 
    if arr[mid]==key:
        return mid 
    elif arr[mid]<key:
        return binary_search(arr,mid+1,right,key)
    else:
        return binary_search(arr,left,mid - 1 , key) 
    

# Example Usage
arr = [1, 3, 5, 7, 9, 11, 13]
target = 70
result = binary_search(arr, 0, len(arr) - 1, target)
if result != -1:
    print(f"Target found at index {result}")
else:
    print("Target not found")